VirtoCommerce.MarketingModule.Web/Scripts/promotion/blades/promotion-expressions.js

```diff
@@ -22,52 +22,165 @@ angular.module('virtoCommerce.marketingModule')
             return;
         }
 
+        var selectedEntry = {};
+        if (!isMultiSelect) {
+            var catalogBlade = {
+                id: "CatalogEntrySelect",
+                title: "marketing.blades.catalog-items-select.title-product",
+                controller: 'virtoCommerce.catalogModule.catalogItemSelectController',
+                template: 'Modules/$(VirtoCommerce.Catalog)/Scripts/blades/common/catalog-items-select.tpl.html',
+                breadcrumbs: [],
+                toolbarCommands: [
+                {
+                    name: "platform.commands.pick-selected", icon: 'fa fa-plus',
+                    executeMethod: function (blade) {
+                        parentElement.productId = selectedEntry.id;
+                        parentElement.productName = selectedEntry.name;
+                        parentElement.productCode = selectedEntry.code;
+                        bladeNavigationService.closeBlade(blade);
+                    },
+                    canExecuteMethod: function () {
+                        return Object.keys(selectedEntry).length > 0;
+                    }
+                }]
+            };
+
+            catalogBlade.options = {
+                showCheckingMultiple: false,
+                checkItemFn: function (listItem, isSelected) {
+                    if (listItem.type == 'category') {
+                        catalogBlade.error = 'Must select Product';
+                        listItem.selected = undefined;
+                    } else {
+                        if (isSelected) {
+                            selectedEntry = listItem;
+                        }
+                        catalogBlade.error = undefined;
+                    }
+                }
+            };
+
+            bladeNavigationService.showBlade(catalogBlade, $scope.blade);
+            return;
+        }
+
+        var openCatalogBlade = function (catalogConditionsBlade, parentBlade) {
+            var catalogBlade = {
+                id: "CatalogEntrySelect",
+                title: "marketing.blades.catalog-items-select.title-product",
+                controller: 'virtoCommerce.catalogModule.catalogItemSelectController',
+                template: 'Modules/$(VirtoCommerce.Catalog)/Scripts/blades/common/catalog-items-select.tpl.html',
+                breadcrumbs: [],
+                toolbarCommands: [
+                {
+                    name: "platform.commands.pick-selected", icon: 'fa fa-plus',
+                    executeMethod: function (blade) {
+                        blade.parentBlade.$scope.gridApi.grid.options.data = _.map(blade.parentBlade.selectedListEntries, function (entry) {
```

Comment:
Probably better to make all work with grid in blade were grid is located and created. As i understand it is 'catalog-entry-list' blade responsibility to add, show or delete items in the list. You could place all these methods in this blade.